File: uniconn-node/src/secure/handshake.ts
import { MlKem1024 } from "mlkem";
import { blake3 } from "@noble/hashes/blake3.js";

import type { IConn } from "../conn.js";
import {
  Identity,
  computeFingerprint,
  verifyWithKey,
  importRawPublicKey,
  type Fingerprint,
} from "./identity.js";
import {
  MSG_HELLO,
  MSG_HELLO_REPLY,
  KDF_CONTEXT,
  KDF_OUTPUT_SIZE,
  NONCE_PREFIX_SIZE,
} from "./constants.js";
import {
  marshalHello,
  unmarshalHello,
  readFrame,
  writeFrame,
  type HelloMsg,
} from "./message.js";
import { SecureConn, type SessionKeys } from "./conn.js";

/** Derive session keys from ML-KEM shared secret using BLAKE3. */
function deriveKeys(sharedSecret: Uint8Array): SessionKeys {
  const ctxBytes = new TextEncoder().encode(KDF_CONTEXT);
  const combined = new Uint8Array(ctxBytes.length + sharedSecret.length);
  combined.set(ctxBytes);
  combined.set(sharedSecret, ctxBytes.length);
  const out = blake3(combined, { dkLen: KDF_OUTPUT_SIZE });

  return {
    initiatorKey: out.slice(0, 32),
    responderKey: out.slice(32, 64),
    initiatorNoncePfx: out.slice(64, 64 + NONCE_PREFIX_SIZE),
    responderNoncePfx: out.slice(64 + NONCE_PREFIX_SIZE, KDF_OUTPUT_SIZE),
  };
}

const kem = new MlKem1024();

/**
 * Perform the initiator side of the USCP handshake.
 */
export async function handshakeInitiator(
  conn: IConn,
  id: Identity,
  peerFP: Fingerprint,
): Promise<SecureConn> {
  // 1. Ephemeral ML-KEM-1024.
  const [ek, dk] = await kem.generateKeyPair();

  // 2. Sign raw ek.
  const sig = id.sign(ek);

  // 3. Send HELLO.
  const hello: HelloMsg = {
    type: MSG_HELLO,
    publicKey: id.publicKeyRaw,
    payload: ek,
    signature: sig,
  };
  await writeFrame(conn, marshalHello(hello));

  // 4. Receive HELLO_REPLY.
  const replyBody = await readFrame(conn);
  const reply = unmarshalHello(replyBody);
  if (reply.type !== MSG_HELLO_REPLY) {
    throw new Error(
      `expected HELLO_REPLY (0x${MSG_HELLO_REPLY.toString(16)}), got 0x${reply.type.toString(16)}`,
    );
  }

  // 5. Verify peer.
  const gotFP = computeFingerprint(reply.publicKey);
  if (!equalBytes(gotFP, peerFP)) {
    throw new Error("fingerprint mismatch");
  }
  const peerPubKeyObj = importRawPublicKey(reply.publicKey);
  if (!verifyWithKey(reply.signature, reply.payload, peerPubKeyObj)) {
    throw new Error("signature verification failed");
  }

  // 6. Decapsulate.
  const sharedSecret = await kem.decap(reply.payload, dk);

  // 7. Derive keys.
  const keys = deriveKeys(sharedSecret);
  return new SecureConn(conn, keys, true);
}

/**
 * Perform the responder side of the USCP handshake.
 */
export async function handshakeResponder(
  conn: IConn,
  id: Identity,
  peerFP: Fingerprint,
): Promise<SecureConn> {
  // 1. Receive HELLO.
  const helloBody = await readFrame(conn);
  const hello = unmarshalHello(helloBody);
  if (hello.type !== MSG_HELLO) {
    throw new Error(
      `expected HELLO (0x${MSG_HELLO.toString(16)}), got 0x${hello.type.toString(16)}`,
    );
  }

  // 2. Verify peer.
  const gotFP = computeFingerprint(hello.publicKey);
  if (!equalBytes(gotFP, peerFP)) {
    throw new Error("fingerprint mismatch");
  }
  const peerPubKeyObj = importRawPublicKey(hello.publicKey);
  if (!verifyWithKey(hello.signature, hello.payload, peerPubKeyObj)) {
    throw new Error("signature verification failed");
  }

  // 3. Encapsulate.
  const [ct, sharedSecret] = await kem.encap(hello.payload);

  // 4. Sign ct.
  const sig = id.sign(ct);

  // 5. Send HELLO_REPLY.
  const reply: HelloMsg = {
    type: MSG_HELLO_REPLY,
    publicKey: id.publicKeyRaw,
    payload: ct,
    signature: sig,
  };
  await writeFrame(conn, marshalHello(reply));

  // 6. Derive keys.
  const keys = deriveKeys(sharedSecret);
  return new SecureConn(conn, keys, false);
}

/** Constant-time byte comparison. */
function equalBytes(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a[i] ^ b[i];
  }
  return diff === 0;
}

File: uniconn-py/uniconn/secure/handshake.py
"""USCP v1 handshake: initiator and responder."""

from __future__ import annotations

from typing import Callable

import blake3 as _blake3

from ..conn import Conn
from .constants import (
    KDF_CONTEXT,
    KDF_OUTPUT_SIZE,
    MSG_HELLO,
    MSG_HELLO_REPLY,
    NONCE_PREFIX_SIZE,
)
from .identity import Identity, compute_fingerprint, verify as _default_verify
from .message import marshal_hello, read_frame, unmarshal_hello
from .conn import SecureConn

# Types
VerifyFn = Callable[[bytes, bytes, bytes], bool]


class SessionKeys:
    """Session keys derived from KDF."""

    __slots__ = (
        "initiator_key",
        "responder_key",
        "initiator_nonce_pfx",
        "responder_nonce_pfx",
    )

    def __init__(
        self,
        initiator_key: bytes,
        responder_key: bytes,
        initiator_nonce_pfx: bytes,
        responder_nonce_pfx: bytes,
    ) -> None:
        self.initiator_key = initiator_key
        self.responder_key = responder_key
        self.initiator_nonce_pfx = initiator_nonce_pfx
        self.responder_nonce_pfx = responder_nonce_pfx


def _derive_keys(shared_secret: bytes) -> SessionKeys:
    """
    Derive session keys from ML-KEM shared secret using BLAKE3.
    KDF: BLAKE3(KDF_CONTEXT || shared_secret, 88 bytes)
    """
    h = _blake3.blake3(KDF_CONTEXT + shared_secret)
    out = h.digest(length=KDF_OUTPUT_SIZE)

    return SessionKeys(
        initiator_key=out[0:32],
        responder_key=out[32:64],
        initiator_nonce_pfx=out[64:76],
        responder_nonce_pfx=out[76:88],
    )


async def handshake_initiator(
    conn: Conn,
    identity: Identity,
    peer_fp: bytes,
    verify_fn: VerifyFn | None = None,
) -> SecureConn:
    """
    Perform the initiator side of the USCP handshake.

    Args:
        conn: Underlying connection (TCP, WebSocket, etc.)
        identity: Local ML-DSA-87 identity.
        peer_fp: Expected 64-byte fingerprint of the peer.
        verify_fn: Optional custom verify function. Defaults to pqcrypto.
    """
    if verify_fn is None:
        verify_fn = _default_verify

    # 1. Ephemeral ML-KEM-1024.
    try:
        from pqcrypto.kem.kyber1024 import generate_keypair, decrypt
    except ImportError:
        raise RuntimeError("pqcrypto package required for ML-KEM-1024")

    pk_kem, sk_kem = generate_keypair()
    ek_bytes = bytes(pk_kem)

    # 2. Sign ek.
    sig = identity.sign(ek_bytes)

    # 3. Send HELLO.
    hello = marshal_hello(MSG_HELLO, identity.public_key_bytes(), ek_bytes, sig)
    await conn.write(hello)

    # 4. Receive HELLO_REPLY.
    reply_body = await read_frame(conn)
    reply = unmarshal_hello(reply_body)
    if reply["msg_type"] != MSG_HELLO_REPLY:
        raise ValueError(
            f"expected HELLO_REPLY (0x{MSG_HELLO_REPLY:02x}), "
            f"got 0x{reply['msg_type']:02x}"
        )

    # 5. Verify peer.
    got_fp = compute_fingerprint(reply["public_key"])
    if got_fp != peer_fp:
        raise ValueError("fingerprint mismatch")
    if not verify_fn(reply["public_key"], reply["payload"], reply["signature"]):
        raise ValueError("signature verification failed")

    # 6. Decapsulate.
    shared_secret = bytes(decrypt(sk_kem, reply["payload"]))

    # 7. Derive keys.
    keys = _derive_keys(shared_secret)
    return SecureConn(keys, is_initiator=True)


async def handshake_responder(
    conn: Conn,
    identity: Identity,
    peer_fp: bytes,
    verify_fn: VerifyFn | None = None,
) -> SecureConn:
    """
    Perform the responder side of the USCP handshake.

    Args:
        conn: Underlying connection.
        identity: Local ML-DSA-87 identity.
        peer_fp: Expected 64-byte fingerprint of the peer.
        verify_fn: Optional custom verify function.
    """
    if verify_fn is None:
        verify_fn = _default_verify

    # 1. Receive HELLO.
    hello_body = await read_frame(conn)
    hello = unmarshal_hello(hello_body)
    if hello["msg_type"] != MSG_HELLO:
        raise ValueError(
            f"expected HELLO (0x{MSG_HELLO:02x}), got 0x{hello['msg_type']:02x}"
        )

    # 2. Verify peer.
    got_fp = compute_fingerprint(hello["public_key"])
    if got_fp != peer_fp:
        raise ValueError("fingerprint mismatch")
    if not verify_fn(hello["public_key"], hello["payload"], hello["signature"]):
        raise ValueError("signature verification failed")

    # 3. Encapsulate.
    try:
        from pqcrypto.kem.kyber1024 import encrypt
    except ImportError:
        raise RuntimeError("pqcrypto package required for ML-KEM-1024")

    ct, shared_secret = encrypt(hello["payload"])
    ct_bytes = bytes(ct)
    shared_secret = bytes(shared_secret)

    # 4. Sign ct.
    sig = identity.sign(ct_bytes)

    # 5. Send HELLO_REPLY.
    reply = marshal_hello(
        MSG_HELLO_REPLY, identity.public_key_bytes(), ct_bytes, sig
    )
    await conn.write(reply)

    # 6. Derive keys.
    keys = _derive_keys(shared_secret)
    return SecureConn(keys, is_initiator=False)
